### ISSUE TITLE
laravel-websockets with Let's Encrypt

### ISSUE THREAD

**Issue by kerrsmith** [beyondcode/laravel-websockets#52]
I have got the code working without SSL and it works as expected. I changed the code as follows to use my Let's Encrypt certificates but can not get it to work:

(I am running the code on a Debian 8 (Jessie) machine configured using ISPConfig)

added encrypted: true, to the Echo code:

window.Echo = new Echo({
    broadcaster: 'pusher',
    key: process.env.MIX_PUSHER_APP_KEY,
    //cluster: process.env.MIX_PUSHER_APP_CLUSTER,
    wsHost: window.location.hostname,
    wsPort: 6001,
    disableStats: true,
    encrypted: true,
});

In the websockets.php file I added my certificates:

 'ssl' => [
        /*
         * Path to local certificate file on filesystem. It must be a PEM encoded file which
         * contains your certificate and private key. It can optionally contain the
         * certificate chain of issuers. The private key also may be contained
         * in a separate file specified by local_pk.
         */
        'local_cert' => '/var/www/example.com/ssl/example.com-le.crt',
        /*
         * Path to local private key file on filesystem in case of separate files for
         * certificate (local_cert) and private key.
         */
        'local_pk' => '/var/www/example.com/ssl/example.com-le.key',
        /*
         * Passphrase for your local_cert file.
         */
        'passphrase' => null,
    ],

When I try to connect in Firefox (or Chrome) I get the following errors:

WebSocket connection to 'wss://example.com/app/f5d5d4fce53eec8a4f05?protocol=7&client=js&version=4.3.1&flash=false' failed: Error during WebSocket handshake: Unexpected response code: 404

then after this:

app.js:39131 WebSocket connection to 'wss://example.com/app/f5d5d4fce53eec8a4f05?protocol=7&client=js&version=4.3.1&flash=false' failed: WebSocket is closed before the connection is established.

My certificates are working as the page loads fine, it's just Echo does not seem to want to connect.

Is there a way to get more information from php artisan websockets:serve? I have tried adding --verbose at the end and it still shows 'Starting the WebSocket server on port 6001..' as normal. Even if I deliberately misspell the certificate name it shows the same message.

**Comment by kerrsmith** [comment]
After a bit of searching around I have now got it working - I did the following:

The Let's Encrypt certificates used by ISPConfig are only readable by root so I tested them by just copying them to my users home folder.

I then followed the setup guide for the 'Usage with Laravel Valet', this meant adding:

wssPort: 6001, to the Echo code:

window.Echo = new Echo({
    broadcaster: 'pusher',
    key: process.env.MIX_PUSHER_APP_KEY,
    //cluster: process.env.MIX_PUSHER_APP_CLUSTER,
    //encrypted: true,
    wsHost: window.location.hostname,
    wsPort: 6001,
    wssPort: 6001,
    disableStats: true
});

and 

'scheme' => 'https' as well as the curl options to broadcasting.php

        'pusher' => [
            'driver' => 'pusher',
            'key' => env('PUSHER_APP_KEY'),
            'secret' => env('PUSHER_APP_SECRET'),
            'app_id' => env('PUSHER_APP_ID'),
            'options' => [
                'cluster' => env('PUSHER_APP_CLUSTER'),
                //'encrypted' => true,
                'host' => '127.0.0.1',
                'port' => 6001,
                'scheme' => 'https',
                'curl_options' => [
                    CURLOPT_SSL_VERIFYHOST => 0,
                    CURLOPT_SSL_VERIFYPEER => 0,
                ]
            ],
        ],

Next in websockets.php I changed by certificate locations and added in 'verify_peer' => false,

'ssl' => [
/*
* Path to local certificate file on filesystem. It must be a PEM encoded file which
* contains your certificate and private key. It can optionally contain the
* certificate chain of issuers. The private key also may be contained
* in a separate file specified by local_pk.
/
'local_cert' => '/my/home/directory/ssl/example.com-le.crt',
/
* Path to local private key file on filesystem in case of separate files for
* certificate (local_cert) and private key.
/
'local_pk' => '/my/home/directory/ssl/example.com-le.key',
/
* Passphrase for your local_cert file.
*/
'passphrase' => null,
'verify_peer' => false,
],

Once this was done it all started working in both Firefox and Chrome.  Basically it was failing because it could not read my certificates as they were not accessible to my webserver user.

**Comment by kerrsmith** [comment]
This could mean the port you are trying to connect to is closed. Does the connection work when you are not using SSL? If it does then it must be something to do with the SSL setup - have you followed the setup for SSL on the following page:

https://docs.beyondco.de/laravel-websockets/1.0/basic-usage/ssl.html#usage-with-laravel-valet

Even if you are not using Laravel Valet this setup is a good place to start, you need to do everything it says as some parts make it work in both Firefox and Chrome.

Have you checked that the port is actually open in Debian:

lsof -nPi tcp:6001

this command should output something similar to the following:

php     12832 webuser    6u  IPv4 27154618      0t0  TCP *:6001 (LISTEN)

I would double check the above as a start and see if everything looks correct.

**Comment by lsfiege** [comment]
Yes, it works without ssl and the port is open. Even the result of lsof is similar to the one you have indicated to me. It seems to be a configuration of letsencrypt / ssl on my server, although I have not yet found which one is

**Comment by lsfiege** [comment]
Finally, I have done the same thing in another VPS and it works correctly. The problem here is my Debian. Sorry

**Comment by lsfiege** [comment]
I have come back here because maybe someone knows that this problem may be caused. When executing the `php artisan websockets:serve` command, and checking on the other hand if indeed this command works on port 6001, the `lsof -nPi tcp:6001` command returns nothing. I have investigated a little about what could be due, but without success. I have also verified that the port is open running `netstat -pln`  
`tcp        0      0 0.0.0.0:6001            0.0.0.0:*               LISTEN      -`
